mgc/discriminability/base.py

```diff
@@ -0,0 +1,123 @@
+from abc import ABC, abstractmethod
+from sklearn.metrics import euclidean_distances
+from sklearn.utils import check_X_y
+import numpy as np
+
+
+class DiscriminabilityTest(ABC):
+    r"""
+    A base class for a Discriminability test.
+    
+    """
+    
+    
+    def __init__(self):
+        super().__init__()
+
+
+    def _statistic(self, X, Y, isDist = False, remove_isolates=True, return_rdfs=False):
+        r"""
+        Calulates the independence test statistic.
+            
+        Parameters
+        ----------
+        x, y : ndarray
+        Input data matrices.
+        """
+
+        uniques, counts = np.unique(Y, return_counts=True)
+        
+        if remove_isolates:
+            idx = np.isin(Y, uniques[counts != 1])
+            labels = Y[idx]
+            
+            if ~isDist:
+                X = X[idx]
+            else:
+                X = X[np.ix_(idx, idx)]
+        else:
+            labels = Y
+
+        if ~isDist:
+            dissimilarities = euclidean_distances(X)
+        else:
+            dissimilarities = X
+
+        rdfs = self._discr_rdf(dissimilarities, labels)
+        stat = np.nanmean(rdfs)
+
+        if return_rdfs:
+            return stat, rdfs
+        else:
+            return stat
+
+
+
+
+    def _discr_rdf(self, dissimilarities, labels):
+    
+        check_X_y(dissimilarities, labels, accept_sparse=True)
+
+        rdfs = []
+        for i, label in enumerate(labels):
+            di = dissimilarities[i]
+
+            # All other samples except its own label
+            idx = labels == label
```

Comment:
Again, using np.unique instead of the next 4 lines, there is likely a better of way of doing this one 1 line